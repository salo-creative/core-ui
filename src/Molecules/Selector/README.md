# Selector

The selector component is for use in place of a standard form select element for areas that concern navigation. It should not be used in forms as a replacement for a true select. If this is your use case please use the Select component in the Forms section.

## Usage

Install

```javascript
yarn add @salo/core-ui
```

Include

```javascript
import Selector from '@salo/core-ui/Molecules/Selector';
```

Implement as follows

```javascript
const { value } = this.state;

<Selector 
  data={ [
    { value: '1', label: 'Option 1' },
    { value: '2', label: 'Option 2' },
    { value: '3', label: 'Option 3' }
  ] }
  value={ value }
  onChange={ (value) => this.setState({ value }) }
/>
```