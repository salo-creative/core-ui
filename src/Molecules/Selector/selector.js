import React from 'react';
import PropTypes from 'prop-types';
import { find, get, isEmpty } from 'lodash';

// COMPONENTS
import { colours } from '../../helpers/colours';
import { IconAfter, Label } from './selector.styles';
import Drop from '../Drop';
import Button from '../Button';

class Selector extends React.Component {
  constructor() {
    super();
    this.state = {
      dropOpen: false,
      element: null
    };
  }
  
  handleChange = (e, value) => {
    const { onChange } = this.props;
    this.setState({ dropOpen: false });
    if (typeof onChange === 'function') {
      onChange(value, e);
    }
  }

  
  dropClose = (e) => {
    const { element } = this.state;
    if (e.target !== element) {
      this.setState({ dropOpen: false });
    }
  };
  
  openDrop = (e) => {
    e.preventDefault();
    this.setState({ element: e.target, dropOpen: true });
  }
  
  renderItems() {
    const { data, valueKey, labelKey, size } = this.props;
    if (isEmpty(data)) return null;
    return data.map(item => (
      <Button
        key={ item[valueKey] }
        onClick={ (e) => this.handleChange(e, item[valueKey]) }
        appearance='text'
        fullWidth
        radius={ false }
        align='start'
        size={ size }
      >
        { item[labelKey] }
      </Button>
    ));
  }

  renderDropDown() {
    const { fixedPositionDrop, dropDownBackground, dropBorder } = this.props;
    const { dropOpen, element } = this.state;
    const width = get(element, 'clientWidth', 0);
    return (
      <Drop
        background={ dropDownBackground }
        width={ width + 2 || 0 }
        disableOverflow={ false }
        element={ element }
        onClose={ (e) => this.dropClose(e) }
        top={ fixedPositionDrop ? 70 : 0 }
        offsetTop={ !fixedPositionDrop ? 2 : 0 }
        offsetLeft={ 1 }
        fixed={ fixedPositionDrop }
        open={ dropOpen }
        border={ dropBorder ? `1px solid ${ colours.blue }` : 'none' }
        shadow='none'
        spacing={ 0 }
        zIndex={ 100 }
      >
        { this.renderItems() }
      </Drop>
    );
  }

  renderIconAfter() {
    const { iconAfter } = this.props;
    if (iconAfter) {
      return (
        <IconAfter>
          <svg style={ { width: '24px', height: '24px' } } viewBox='0 0 24 24'>
            <path d='M7,10L12,15L17,10H7Z' />
          </svg>
        </IconAfter>
      );
    }
    return null;
  }

  renderValue() {
    const { value, valueKey, placeholder, labelKey, data } = this.props;
    if (!value || !find(data, { [valueKey]: value }, null)) return placeholder;
    return find(data, { [valueKey]: value }, null)[labelKey];
  }

  render() {
    const { value, placeholder, background, size } = this.props;

    const label = this.renderValue();
    return (
      <React.Fragment>
        <Label
          value={ value }
          onClick={ (e) => this.openDrop(e) }
          background={ background }
          className={ label === placeholder ? 'noValue' : '' }
          size={ size }
        >
          { label }
          { this.renderIconAfter() }
        </Label>
        { this.renderDropDown() }
      </React.Fragment>
    );
  }
}

Selector.defaultProps = {
  value: '',
  onChange: null,
  iconAfter: true,
  data: [],
  valueKey: 'value',
  labelKey: 'label',
  placeholder: 'Please selectâ€¦',
  background: 'transparent',
  dropDownBackground: '#fff',
  fixedPositionDrop: false,
  dropBorder: true,
  size: 'M'
};

Selector.propTypes = {
  iconAfter: PropTypes.bool,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  data: PropTypes.array,
  valueKey: PropTypes.string,
  labelKey: PropTypes.string,
  placeholder: PropTypes.string,
  background: PropTypes.string,
  dropDownBackground: PropTypes.string,
  fixedPositionDrop: PropTypes.bool,
  dropBorder: PropTypes.bool,
  size: PropTypes.oneOf(['L', 'M'])
};

export default Selector;