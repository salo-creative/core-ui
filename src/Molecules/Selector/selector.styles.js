import styled from 'styled-components';
import { transparentize } from 'polished';

export const Label = styled.button`
 position: relative;
  margin: 0;
  width: 100%;
  outline: none;
  width: 100%;
  height: ${ ({ size }) => (size === 'L' ? '4.5rem' : '4rem') };
  border-radius: 0.4rem;
  background: transparent;
  color: ${ ({ color }) => color };
  font-size: ${ ({ size }) => (size === 'L' ? '1.6rem' : '1.4rem') };
  padding: 0 4rem 0 1rem;
  border: 1px solid ${ ({ theme }) => transparentize(0.5, theme.primary) };
  margin: 0;
  display: flex;
  align-items: center;
  cursor: pointer;
  &:focus,
  &:active {
    border: 1px solid ${ ({ theme }) => theme.primary };
  }
  &.noValue {
    color: ${ ({ theme }) => theme.darkGrey };
  }
  &:after {
    content:'';
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0;
    left: 0;
  }
`;

export const IconAfter = styled.span`
  display: flex;
  width: 40px;
  height: calc(100% - 10px);
  align-items: center;
  right: 0;
  top: 5px;
  position: absolute;
  justify-content: center;
  border-left: 1px solid ${ ({ theme }) => transparentize(0.5, theme.primary) };
  svg {
    fill: ${ ({ theme }) => theme.darkGrey };
  }
`;