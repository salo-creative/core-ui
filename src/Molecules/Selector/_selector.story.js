import React from 'react';
import { withKnobs, select } from '@storybook/addon-knobs';
import { RenderWithProps, Store } from '@jamesbliss/storybook-state';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENT
import { Selector } from '../../index';

// README //
import README from './README.md';

// Start of story logic
export const Basic = () => {
  const size = select('size', ['L', 'M'], 'M');
  // STORE
  const store = new Store({
    value: ''
  });
  return (
    <RenderWithProps store={ store }>
      <Selector
        data={ [
          {
            value: '1', label: 'Option 1'
          },
          {
            value: '2', label: 'Option 2'
          },
          {
            value: '3', label: 'Option 3'
          }
        ] }
        value={ store.value }
        onChange={ value => store.set({
          value
        }) }
        size={ size }
      />
    </RenderWithProps>
  );
};

Basic.story = {
  decorators: [
    withKnobs,
    withTests({
      results
    })
  ],
  parameters: {
    jest: ['selector'],
    info: {
      propTablesExclude: []
    },
    notes: README
  }
};

export default {
  title: 'Molecules/Selector'
};