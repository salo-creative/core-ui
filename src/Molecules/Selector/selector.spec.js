import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import Selector from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <Selector />
  );
  expect(container.firstChild).toMatchSnapshot();
});