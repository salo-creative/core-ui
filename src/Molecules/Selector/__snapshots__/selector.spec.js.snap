// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c0 {
  position: relative;
  margin: 0;
  width: 100%;
  outline: none;
  width: 100%;
  height: 4rem;
  border-radius: 0.4rem;
  background: transparent;
  font-size: 1.4rem;
  padding: 0 4rem 0 1rem;
  border: 1px solid rgba(0,172,237,0.5);
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c0:focus,
.c0:active {
  border: 1px solid #00ACED;
}

.c0.noValue {
  color: #6c757d;
}

.c0:after {
  content: '';
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 40px;
  height: calc(100% - 10px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  right: 0;
  top: 5px;
  position: absolute;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-left: 1px solid rgba(0,172,237,0.5);
}

.c1 svg {
  fill: #6c757d;
}

<button
  class="noValue c0"
  value=""
>
  Please selectâ€¦
  <span
    class="c1"
  >
    <svg
      style="width: 24px; height: 24px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M7,10L12,15L17,10H7Z"
      />
    </svg>
  </span>
</button>
`;
